File: bionics/speaker.py
import threading
import time
import os
import urllib.parse
import urllib.request
from slugify import slugify
from sys import platform
from bionics.queues import Queues

class Speaker (threading.Thread):

    def __init__(self):
        threading.Thread.__init__(self)

    def run(self):
        while True:

            if not Queues.message.empty():

                message = Queues.message.get()
                error = 0

                sound_directory = 'sounds'
                os.makedirs(sound_directory, exist_ok=True)

                file_path = sound_directory + '/' + slugify(message) + '.wav'

                if not (os.path.exists(file_path)):
                    params = (('INPUT_TEXT', message),
                              ('INPUT_TYPE', 'TEXT'),
                              ('OUTPUT_TYPE', 'AUDIO'),
                              ('AUDIO', 'WAVE_FILE'),
                              ('LOCALE', 'de'),
                              ('effect_Volume_selected', 'on'),
                              ('effect_Volume_parameters', 'amount:2.0'),
                              ('effect_Chorus_selected', 'on'),
                              ('effect_Chorus_parameters',
                               'delay1:466;amp1:0.54;delay2:600;amp2:-0.10;delay3:250;amp3:0.30'))

                    try:
                        urllib.request.urlretrieve('http://mary.dfki.de:59125/process?' + urllib.parse.urlencode(params), file_path)

                    except:
                        error = 1

                if not error:

                    Queues.command.put(message)

                    if 'darwin' == platform:
                        os.system('afplay ' + file_path + ' > /dev/null 2>&1')
                    else:
                        os.system('mplayer ' + file_path + ' > /dev/null 2>&1')

            else:
                time.sleep(0.1)

File: bionics/speaker.js
const config = require('./config').speaker;
const http = require('http');
const logger = require('./logger');
const spawn = require('child_process').spawn;
const fs = require('node-fs');
const querystring = require('querystring');
const slug = require('slug');
const player = require('play-sound')(opts = {})


class Speaker {

    constructor() {
        this._ready = false;
        this._working = false;
        this._queue = [];

        if (!fs.existsSync(config.directory)) fs.mkdirSync(config.directory);

        http.get('http://' + config.marytts.host + ':' + config.marytts.port + '/version', (response) => {

            if(response && 200 == response.statusCode) {
                this._ready = true;
            } else {
                this._startMaryTTS();
            }

        }).on('error', (e) => {
            this._startMaryTTS();
        });

    }

    get ready() {
        return this._ready;
    }

    _startMaryTTS() {

        logger.info('Start MaryTTS Server');

        const child = spawn('./' + config.marytts.path, {
            detached: true
        });

        child.stdout.on('data', (data) => {
            logger.debug(data.toString());
        });

        child.stderr.on('data', (data) => {
            if(data.toString().includes('started in') && data.toString().includes('on port')) this._ready = true;
            logger.debug(data.toString());
        });

        child.on('close', (code) => {
            logger.debug('child process exited with code: ' + code);
        });

        child.unref();

    }

    say(message) {

        // queue message if current say process running
        if(this._working) {
            this._queue.push(message);
            return false;
        }

        this._working = true;

        let params = {
            'INPUT_TEXT' : message,
            'INPUT_TYPE': 'TEXT',
            'OUTPUT_TYPE' : 'AUDIO',
            'AUDIO' : 'WAVE_FILE',
            'LOCALE' : 'de',
            'effect_Chorus_selected' : 'on',
            'effect_Chorus_parameters' : 'delay1:466;amp1:0.54;delay2:600;amp2:-0.10;delay3:250;amp3:0.30'
        };

        let queryString = querystring.stringify(params);
        let url = 'http://' + config.marytts.host + ':' + config.marytts.port + '/process?' + queryString;
        let filePath = config.directory + '/' + slug(message, {lower: true}) + '.wav';
        let errorMessage = 'can not get message from marytts server for url: ' + url;

        http.get(url, (response) => {

            if(response && 200 === response.statusCode) {

                let file = fs.createWriteStream(filePath);

                file.on('finish', () => {
                    file.close(() => {
                        player.play(filePath, (error) => {
                            if (error) {
                                logger.error(error);
                                throw error;
                            } else {
                                logger.debug(message);
                                this._working = false;

                                if(this._queue.length > 0) {
                                    this.say(this._queue.shift());
                                }
                            }
                        })
                    });
                });

                response.pipe(file);

            } else {
                logger.error(errorMessage);
                throw errorMessage;
            }

        }).on('error', (e) => {
            logger.error(errorMessage);
            throw errorMessage;
        });

    }
}

module.exports = new Speaker();